node/src/store/in_memory_store.rs

```diff
@@ -1,199 +1,306 @@
+        self.inner.read().await.get_stored_ranges()
     }
 
-    fn get_head(&self) -> Result<ExtendedHeader> {
-        let head_height = self.get_head_height()?;
-        self.get_by_height(head_height)
+    /// Clone the store and all its contents. Async fn due to internal use of async mutex.
+    pub async fn async_clone(&self) -> Self {
+        InMemoryStore {
+            inner: RwLock::new(self.inner.read().await.clone()),
+            sampling_data: self.sampling_data.clone(),
+            header_added_notifier: Notify::new(),
+        }
+    }
+}
+
+impl InMemoryStoreInner {
+    fn get_stored_ranges(&self) -> HeaderRanges {
+        self.stored_ranges.clone()
+    }
+
+    #[inline]
+    fn get_head_height(&self) -> Result<u64> {
+        self.stored_ranges.head().ok_or(StoreError::NotFound)
     }
 
     fn contains_hash(&self, hash: &Hash) -> bool {
         self.headers.contains_key(hash)
     }
 
     fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader> {
-        self.headers
-            .get(hash)
-            .as_deref()
-            .cloned()
-            .ok_or(StoreError::NotFound)
+        self.headers.get(hash).cloned().ok_or(StoreError::NotFound)
     }
 
     fn contains_height(&self, height: u64) -> bool {
-        let Ok(head_height) = self.get_head_height() else {
-            return false;
-        };
-
-        height != 0 && height <= head_height
+        self.stored_ranges.contains(height)
     }
 
     fn get_by_height(&self, height: u64) -> Result<ExtendedHeader> {
-        if !self.contains_height(height) {
+        let Some(hash) = self.height_to_hash.get(&height).copied() else {
             return Err(StoreError::NotFound);
-        }
-
-        let Some(hash) = self.height_to_hash.get(&height).as_deref().copied() else {
-            return Err(StoreError::LostHeight(height));
         };
 
         self.headers
             .get(&hash)
-            .as_deref()
             .cloned()
             .ok_or(StoreError::LostHash(hash))
     }
 
-    fn update_sampling_metadata(
-        &self,
-        height: u64,
-        status: SamplingStatus,
-        cids: Vec<Cid>,
+    async fn insert(
+        &mut self,
+        headers: Vec<ExtendedHeader>,
+        verify_neighbours: bool,
     ) -> Result<()> {
-        if !self.contains_height(height) {
-            return Err(StoreError::NotFound);
+        let (Some(head), Some(tail)) = (headers.first(), headers.last()) else {
+            return Ok(());
+        };
+
+        let headers_range = head.height().value()..=tail.height().value();
+        let range_scan_result = self.stored_ranges.check_range_insert(&headers_range)?;
+
+        if verify_neighbours {
+            let prev_exists = headers_range.start() != range_scan_result.range.start();
+            let next_exists = headers_range.end() != range_scan_result.range.end();
+            // header range is already internally verified against itself in `P2p::get_unverified_header_ranges`
```

Comment:
That is true, however this is a public API and we shouldn't trust the input. So we have 2 options:

1. Add `verify_headers` flag.
2. Make `insert` a private API.

This issue exists in other stores too.

In case of option 1, I prefer something like:

```rust
enum Verify {
    No,
    Headers,
    Neightbours,
    HeadersAndNeightbours,
}
```

Reply:
I've put it here mostly to have a similar API to what we had previously with `append_single`/`append_single_unchecked`. If we don't want to provide unchecked part, and enforce checks, I think we could remove it entirely.